File: integrations/sdk/python/bais_client.py
"""
BAIS Python SDK Client

Simple client for integrating BAIS governance into any Python application.

Usage:
    from bais_client import BAISClient
    
    client = BAISClient(api_url="https://api.bais.invitas.ai")
    
    # Audit a response
    result = client.audit(
        query="What is 2+2?",
        response="The answer is 4.",
        domain="general"
    )
    
    if result.approved:
        print("Response approved")
    else:
        print(f"Issues: {result.issues}")
"""

import requests
from dataclasses import dataclass
from typing import List, Dict, Any, Optional


@dataclass
class AuditResult:
    """Result of a BAIS audit."""
    approved: bool
    confidence: float
    issues: List[str]
    warnings: List[str]
    recommendation: str
    should_regenerate: bool
    improved_response: Optional[str] = None


@dataclass
class VerificationResult:
    """Result of a completion verification."""
    valid: bool
    confidence: float
    violations: List[str]
    clinical_status: str  # truly_working, incomplete, stubbed, simulated, fallback, failover


class BAISClient:
    """
    BAIS Governance Client
    
    Integrates BAIS cognitive governance into any application.
    """
    
    def __init__(
        self,
        api_url: str = "http://localhost:8000",
        api_key: Optional[str] = None,
        timeout: int = 30
    ):
        """
        Initialize BAIS client.
        
        Args:
            api_url: Base URL of BAIS API
            api_key: Optional API key for authentication
            timeout: Request timeout in seconds
        """
        self.api_url = api_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self._session = requests.Session()
        
        if api_key:
            self._session.headers["Authorization"] = f"Bearer {api_key}"
    
    def audit(
        self,
        query: str,
        response: str,
        domain: str = "general",
        documents: Optional[List[Dict]] = None
    ) -> AuditResult:
        """
        Audit an LLM response for bias, errors, and quality issues.
        
        Args:
            query: The user's original query
            response: The LLM's response to audit
            domain: Domain context (general, medical, financial, legal)
            documents: Optional source documents for grounding verification
        
        Returns:
            AuditResult with approval status and any detected issues
        """
        payload = {
            "query": query,
            "response": response,
            "domain": domain,
            "documents": documents or []
        }
        
        result = self._post("/governance/audit", payload)
        
        return AuditResult(
            approved=result.get("decision") == "approved",
            confidence=result.get("accuracy", 0) / 100,
            issues=result.get("issues", []),
            warnings=result.get("warnings", []),
            recommendation=result.get("recommendation", ""),
            should_regenerate=result.get("should_regenerate", False),
            improved_response=result.get("improved_response")
        )
    
    def verify_completion(
        self,
        claim: str,
        evidence: List[str]
    ) -> VerificationResult:
        """
        Verify a completion claim against evidence.
        
        Args:
            claim: The claim to verify (e.g., "Feature X is complete")
            evidence: List of evidence items supporting the claim
        
        Returns:
            VerificationResult with validity and clinical status
        """
        payload = {
            "claim": claim,
            "evidence": evidence
        }
        
        result = self._post("/governance/verify", payload)
        
        return VerificationResult(
            valid=result.get("valid", False),
            confidence=result.get("confidence", 0),
            violations=result.get("violations", []),
            clinical_status=result.get("clinical_status", "unknown")
        )
    
    def check_query(self, query: str) -> Dict[str, Any]:
        """
        Pre-check a query for manipulation or injection attempts.
        
        Args:
            query: The user query to check
        
        Returns:
            Dict with safe, risk_level, and any issues
        """
        payload = {"query": query}
        return self._post("/governance/check_query", payload)
    
    def improve_response(
        self,
        response: str,
        issues: List[str]
    ) -> str:
        """
        Improve a response based on detected issues.
        
        Args:
            response: The response to improve
            issues: List of issues to address
        
        Returns:
            Improved response text
        """
        payload = {
            "response": response,
            "issues": issues
        }
        
        result = self._post("/governance/improve", payload)
        return result.get("improved_response", response)
    
    def get_statistics(self) -> Dict[str, Any]:
        """Get governance statistics."""
        return self._get("/governance/statistics")
    
    def _post(self, endpoint: str, payload: Dict) -> Dict:
        """Make POST request."""
        url = f"{self.api_url}{endpoint}"
        response = self._session.post(url, json=payload, timeout=self.timeout)
        response.raise_for_status()
        return response.json()
    
    def _get(self, endpoint: str) -> Dict:
        """Make GET request."""
        url = f"{self.api_url}{endpoint}"
        response = self._session.get(url, timeout=self.timeout)
        response.raise_for_status()
        return response.json()


# Convenience functions
def audit(query: str, response: str, **kwargs) -> AuditResult:
    """Quick audit using default client."""
    client = BAISClient()
    return client.audit(query, response, **kwargs)


def verify(claim: str, evidence: List[str]) -> VerificationResult:
    """Quick verification using default client."""
    client = BAISClient()
    return client.verify_completion(claim, evidence)

File: integrations/sdk/javascript/bais-client.js
/**
 * BAIS JavaScript/TypeScript SDK Client
 * 
 * Simple client for integrating BAIS governance into any JavaScript application.
 * 
 * Usage:
 *   import { BAISClient } from 'bais-client';
 *   
 *   const client = new BAISClient({ apiUrl: 'https://api.bais.invitas.ai' });
 *   
 *   const result = await client.audit({
 *     query: 'What is 2+2?',
 *     response: 'The answer is 4.',
 *     domain: 'general'
 *   });
 *   
 *   if (result.approved) {
 *     console.log('Response approved');
 *   }
 */

class BAISClient {
  /**
   * Initialize BAIS client.
   * 
   * @param {Object} options - Configuration options
   * @param {string} options.apiUrl - Base URL of BAIS API
   * @param {string} [options.apiKey] - Optional API key for authentication
   * @param {number} [options.timeout] - Request timeout in milliseconds (default: 30000)
   */
  constructor(options = {}) {
    this.apiUrl = (options.apiUrl || 'http://localhost:8000').replace(/\/$/, '');
    this.apiKey = options.apiKey || null;
    this.timeout = options.timeout || 30000;
  }

  /**
   * Audit an LLM response for bias, errors, and quality issues.
   * 
   * @param {Object} params - Audit parameters
   * @param {string} params.query - The user's original query
   * @param {string} params.response - The LLM's response to audit
   * @param {string} [params.domain] - Domain context (general, medical, financial, legal)
   * @param {Array} [params.documents] - Optional source documents for grounding verification
   * @returns {Promise<AuditResult>} Audit result with approval status and detected issues
   */
  async audit({ query, response, domain = 'general', documents = [] }) {
    const result = await this._post('/governance/audit', {
      query,
      response,
      domain,
      documents
    });

    return {
      approved: result.decision === 'approved',
      confidence: (result.accuracy || 0) / 100,
      issues: result.issues || [],
      warnings: result.warnings || [],
      recommendation: result.recommendation || '',
      shouldRegenerate: result.should_regenerate || false,
      improvedResponse: result.improved_response || null
    };
  }

  /**
   * Verify a completion claim against evidence.
   * 
   * @param {Object} params - Verification parameters
   * @param {string} params.claim - The claim to verify
   * @param {Array<string>} params.evidence - List of evidence items
   * @returns {Promise<VerificationResult>} Verification result with validity and clinical status
   */
  async verifyCompletion({ claim, evidence }) {
    const result = await this._post('/governance/verify', {
      claim,
      evidence
    });

    return {
      valid: result.valid || false,
      confidence: result.confidence || 0,
      violations: result.violations || [],
      clinicalStatus: result.clinical_status || 'unknown'
    };
  }

  /**
   * Pre-check a query for manipulation or injection attempts.
   * 
   * @param {string} query - The user query to check
   * @returns {Promise<Object>} Result with safe, riskLevel, and any issues
   */
  async checkQuery(query) {
    return await this._post('/governance/check_query', { query });
  }

  /**
   * Improve a response based on detected issues.
   * 
   * @param {Object} params - Improvement parameters
   * @param {string} params.response - The response to improve
   * @param {Array<string>} params.issues - List of issues to address
   * @returns {Promise<string>} Improved response text
   */
  async improveResponse({ response, issues }) {
    const result = await this._post('/governance/improve', {
      response,
      issues
    });
    return result.improved_response || response;
  }

  /**
   * Get governance statistics.
   * 
   * @returns {Promise<Object>} Statistics object
   */
  async getStatistics() {
    return await this._get('/governance/statistics');
  }

  /**
   * Make POST request.
   * @private
   */
  async _post(endpoint, payload) {
    const url = `${this.apiUrl}${endpoint}`;
    const headers = {
      'Content-Type': 'application/json'
    };

    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(payload),
        signal: controller.signal
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      return await response.json();
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Make GET request.
   * @private
   */
  async _get(endpoint) {
    const url = `${this.apiUrl}${endpoint}`;
    const headers = {};

    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method: 'GET',
        headers,
        signal: controller.signal
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      return await response.json();
    } finally {
      clearTimeout(timeoutId);
    }
  }
}

// TypeScript type definitions (as JSDoc)
/**
 * @typedef {Object} AuditResult
 * @property {boolean} approved - Whether the response was approved
 * @property {number} confidence - Confidence score (0-1)
 * @property {string[]} issues - List of detected issues
 * @property {string[]} warnings - List of warnings
 * @property {string} recommendation - Recommendation message
 * @property {boolean} shouldRegenerate - Whether response should be regenerated
 * @property {string|null} improvedResponse - Improved response if available
 */

/**
 * @typedef {Object} VerificationResult
 * @property {boolean} valid - Whether the claim is valid
 * @property {number} confidence - Confidence score (0-1)
 * @property {string[]} violations - List of violations found
 * @property {string} clinicalStatus - Clinical status (truly_working, incomplete, stubbed, etc.)
 */

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { BAISClient };
}

if (typeof window !== 'undefined') {
  window.BAISClient = BAISClient;
}

export { BAISClient };

